Documentation/botr/readytorun-format.md

```diff
@@ -1,15 +1,77 @@
+* Easier deployment: desire to avoid having to deal with potentially hundreds or thousands of assemblies.
+* Side-by-side deployment: a self-contained app doesn't need to care about the location of its dependencies, their proper versions, about sharing them with other apps (including other versions of the same app).
+
+Semantic requirements for the composite R2R format flow from the combination of the basic design principle that the R2R file is a code cache to save jitting time and from the fact that the tighter coupling can be utilized to improve codegen quality.
+
+### R2R file is a code cache
+
+A corollary of this basic principle is that the composite file must maintain full functionality of single-input R2R files w.r.t. things like reflection, dynamic or generics. To achieve this goal the composite R2R file must store all the original MSIL metadata blobs for all input assemblies in the output R2R file. This is a substantial deviation from the previous format that only had one COR header addressed by the COM data directory entry in the COFF header.
+
+To achieve this goal, we propose using two complementary strategies:
+
+* In the composite R2R file, there must be a new table of COR headers and metadata blobs representing the MSIL metadata from all the input assemblies. The table must be indexable by simple assembly name for fast lookup. 
+
+* We propose slightly variating the purpose of the "main" COR header and metadata blob: in composite R2R files, where a "singleton" ECMA-335 metadata blob doesn't make much sense, we'd reuse the "main" metadata for the manifest metadata.
+
+* We may be able to leverage this semantic shift to include some virtual "composite assembly information" in the assembly attributes of the manifest metadata assembly to facilitate versioning and deployment.
+
+**Note:** As of this proposal we're not addressing managed C++ code including both native and managed code. For pre-built native code, it's generally not possible to relocate RVA's as the fixup table information is not rich enough. As a consequence, it's simply not possible to roll multiple managed C++ modules containing native code into a single output R2R PE binary. We might be able to consider some mitigations like allowing a single managed binary containing native code in a composite build but I'm not sure whether they are worth the effort and desirable in general.
+
+### Better codegen in composite R2R files
+
+The perf improvements for composite R2R files can stem from the following facts:
+
+* The composite R2R file forms a tightly-bound version bubble where an individual assembly cannot be accidentally replaced on disk so we don't need to emit any type layout checks.
+* Similarly, the runtime doesn't need to check reference assembly MVID's within the tightly-bound version bubble represented by the composite file.
```

Comment:
This is a new change that @MichalStrehovsky is code reviewing right now in a separate PR https://github.com/dotnet/coreclr/pull/25281. Yes, these should be assembly load-time GUID consistency checks. I'm not saying we mustn't be performing them for combined R2R, I just mentioned I don't see them as necessary because we usually don't take software precautions against tampering with a file - that's what we use other deployment options like security envelopes for - but we normally take software precautions against locating a wrong file. For the check itself, it's probably not super-expensive, but with your 1500 assemblies in a composite file, who knows... :-)